sparsify/sparsegpt.py

```diff
@@ -0,0 +1,255 @@
+# This adapts SparseGPT process: https://github.com/IST-DASLab/sparsegpt
+# E. Frantar et al SparseGPT: Massive Language Models Can Be Accurately Pruned in One-Shot, https://arxiv.org/abs/2301.00774
+# portions copyright by the authors licensed under the Apache License 2.0
+
+
+import gc
+import sys
+import time
+from pathlib import Path
+from typing import Optional
+
+import torch
+from datasets import load_dataset
+
+
+wd = Path(__file__).parent.parent.resolve()
+sys.path.append(str(wd))
+
+from lit_llama import LLaMA, Tokenizer
+from lit_llama.sparsification import SparseGPT
+
+from lit_llama.utils import EmptyInitOnDevice, llama_model_lookup
+
+
+def get_sample_data():
+    traindata = load_dataset(
+        "allenai/c4",
+        "allenai--c4",
+        data_files={"train": "en/c4-train.00000-of-01024.json.gz"},
+        split="train",
+    )
+    # heuristic for the data size?
+    txt = "\n".join(
+        traindata[i]["text"] for i in torch.randperm(len(traindata))[:1000].tolist()
+    )
+    return txt
+
+@torch.no_grad()
+def llama_blockwise_sparsification(
+    model, 
+    sample_inputs, 
+    working_device,
+    *,
+    sparsity=0,
+    prunen=0,
+    prunem=0,
+    
+):
+    
+    print('Getting Inputs for the first block')
+    model.transformer.wte.to(working_device)
+    sample_inputs = sample_inputs.to(working_device)
+    inps = model.transformer.wte(sample_inputs)
+    model.transformer.wte.to("cpu")
+    torch.cuda.empty_cache()
+
+    rope_cache = model.build_rope_cache(sample_inputs)
+    mask_cache = model.build_mask_cache(sample_inputs)
+
+    print('Starting to sparsify block')
+    outs = torch.zeros_like(inps)
+
+
+    submodules_to_process = [
+        "attn.c_attn",
+        "attn.c_proj",
+        "mlp.c_fc1",
+        "mlp.c_fc2",
+        "mlp.c_proj",
+    ]
+
+
+    for i, block in enumerate(model.transformer.h):
+
+        block.to(working_device)
+
+        for name in submodules_to_process:
+            print(i, name, end=" ")
+            t0 = time.perf_counter()
+            print("collecting stats", end=" ")
```

Comment:
```python
print("collecting stats", end=" ", flush=True)
```
you can use the `flush` flag instead of `sys.stdout.flush()`

Reply:
Okay cool, Thanks